sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -36,31 +36,84 @@ import com.digitalasset.daml.lf.transaction.{
 import com.digitalasset.daml.lf.value.Value.ValueArithmeticError
 import com.digitalasset.daml.lf.value.{ContractIdVersion, Value => V}
 
-import scala.annotation.{nowarn, tailrec}
+import scala.annotation.{nowarn, tailrec, unused}
 import scala.collection.immutable.ArraySeq
+import scala.reflect.ClassTag
 import scala.util.control.NonFatal
 
 private[lf] object Speedy {
 
   // These have zero cost when not enabled. But they are not switchable at runtime.
   private val enableInstrumentation: Boolean = false
 
-  final class Metrics(val batchSize: Long) {
+  trait MetricPlugin {
+    type Result
```

Comment:
Using an inner type, rather than a type parameter, simplifies call site usage - i.e. call sites can use just the plugin type, rather than result and plugin type.